src/ttl.ts

```diff
@@ -3,46 +3,123 @@ import { client } from './app';
 
 export async function continuallyRetrieveMessages(): Promise<void> {
   while (true) {
     await retrieveMessages();
   }
 }
 
-async function handleDeletesForMessagesOlderThan14Days(
+async function retrieveMessages(): Promise<void> {
+  // eslint-disable-next-line @typescript-eslint/no-misused-promises
+  for (const item of client.channels.cache) {
+    const channel = item[1];
+    if (!channel.isText() || channel.type === 'DM') {
+      continue;
+    }
+    if (!canGetAndDeleteMessages(channel)) {
+      continue;
+    }
+    if (!lastDeletedMessages[channel.id]) {
+      lastDeletedMessages[channel.id] = channel.id;
+    }
+    try {
+      const messages = await channel.messages.fetch({ after: lastDeletedMessages[channel.id], limit: 100 });
+      const awaitedPromises = await handleDeletesForNonBulkDeletableMessages(channel.guildId, channel.id, messages);
+      if (awaitedPromises.length === 0) {
+        await handleDeletesForBulkDeletableMessages(channel.guildId, channel, messages);
+      }
+    } catch (err) {
+      console.error(err);
+    }
+  }
+}
+
+/**
+ * This is a cache so we don't have to keep checking channel permissions all the time.
+ * Channel ID -> Timestamp of when it was added to the dict
+ */
+const channelsWithoutProperPermissions: { [channelId: string]: number } = {};
+const channelPermissionsCacheDuration: number = 1000 * 60 * 60; // 1 hour in millis
+
+function clearOutdatedChannelPermissionsCache() {
+  // eslint-disable-next-line guard-for-in
+  for (const channelId in channelsWithoutProperPermissions) {
+    const timestampOfInsert = channelsWithoutProperPermissions[channelId];
+    if (timestampOfInsert < Date.now() - channelPermissionsCacheDuration) {
+      delete channelsWithoutProperPermissions[channelId];
+    }
+  }
+}
+
+function hasProperPermissionsToGetAndDelete(channel: GuildTextBasedChannel): boolean {
+  const me = channel.guild.me;
+  if (!me) {
+    return false;
+  }
+  const perms = me.permissionsIn(channel);
+  return perms.has('VIEW_CHANNEL') && perms.has('READ_MESSAGE_HISTORY') && perms.has('MANAGE_MESSAGES');
+}
+
+/**
+ * Wrapper function for checking channel permissions non-excessively (by using a cache)
+ */
+function canGetAndDeleteMessages(channel: GuildTextBasedChannel): boolean {
+  clearOutdatedChannelPermissionsCache();
+  if (channelsWithoutProperPermissions[channel.id]) {
+    return false;
+  }
+  if (!hasProperPermissionsToGetAndDelete(channel)) {
+    channelsWithoutProperPermissions[channel.id] = Date.now();
+    return false;
+  }
+  return true;
+}
+
+function isMessageOlderThanTtl(serverId: string, message: { createdAt: { getTime: () => number } }): boolean {
+  return message.createdAt.getTime() < Date.now() - getServerMessageTtlMillis(serverId);
+}
+
+function canMessageBeBulkDeleted(message: { createdAt: { getTime: () => number } }): boolean {
+  const bulkDeletionThresholdInMillis: number = 1209600000; // 14 days (Discord's bulk deletion age threshold)
```

Comment:
Maybe to make it easier to read:
```suggestion
  const bulkDeletionThresholdInMillis: number = 1000 * 60 * 60 * 24 * 14; // 14 days (Discord's bulk deletion age threshold)
```